<h1>Introduction to GNU Make and Build Managers</h1>

<h1>Joanna Malukiewicz</h1>

<h1>AGAR 2018</h1>

<h2>1. Uses for GNU Make</h2>

<p>Although Make is typically used to compile programs written in a language like C, it is also very handy for bioinformatics, data management and data analysis!</p>

<p>In bioinformatics and genomics, a typical pipeline could include steps to pre-treat and filter your sequencer files, perform <em>de novo</em> or referenced genome alignment, further filtering, merging, variant calling, filtering, and then some type of analyses. </p>

<p>You could write a series of shell scripts (which is fine), but Make and other build managers such as Snakemake and Scons give you much finer grain control and organization over running your pipeline than straight shell scripting. These build managers will let you </p>

<ul>
<li><p>Create a Makefile that describes how files in your pipeline depend on each other, and how to update out-of-date files.</p></li>
<li><p>Give flow between file dependencies and steps within your pipeline. </p></li>
<li><p>Reduce redundancy to streamline your coding  (which will make your life much easier in the long run).</p></li>
</ul>

<h2>2. Example 1 (Simple)</h2>

<ul>
<li>Task:
<ul>
<li>Align a number of fastq Illumina files to the human reference genome with BWA </li>
</ul></li>
</ul>



<pre><code>$bwa mem sample1.F1.fastq sample2.R1.fastq &gt; sample1.sam
</code></pre>



<p>What is you had 10s or 100s files to process? This is where GNu Make can help. 
Start by setting up a <em>rule</em> to process your task:</p>



<pre><code>sample1.sam: sample1.F1.fastq sample2.R1.fastq
bwa mem sample1.F1.fastq sample2.R1.fastq &gt; sample1.sam
</code></pre>



<p>sample1.sam is the <em>target</em> (or end result) of your rule 
the creation of the target is dependent on your input files, sample1.F1.fastq and sample2.R1.fastq, or your <em>dependencies</em> . 
The second line, which always has to be indented, is called the <em>action line</em> and contains the actual line of command. </p>

<p>How do you modify the target rule to accomodate several fastq files? </p>

<p>First we need to define a few variables </p>



<pre><code># make a list of all the sam file targets you want to produce and define it as a variable 
DATABASE = output_files.txt 

# define the path to your reference genome as a variable 
REF= /path/to/ref/genome/hg38.fa

#define a variable that will act as a shell command that will read your list of output files via the shell cat command 
OUTPUTFILES := $(shell cat $(DATABASE))
</code></pre>



<p>Then set targets so that Make will process your entire list of output targets. </p>

<p>To get Make to build everything at once, we have to introduce a phony target, which doesn't correspond to any files. </p>

<p>Since a phony target doesn't actually exist, it won't ever be updated</p>



<pre><code>all: $(OUTPUTFILES)
</code></pre>



<p>$(VARIALBLE_NAME) is a way to call whatever variable you set within a Makefile </p>

<p>Then finally the generalized rule will use some automatic Make variables. </p>



<pre><code>%.bam  :  %.unassembled.forward.fastq /%.unassembled.reverse.fastq
    bwa mem -t 10 $(REF) $^ &gt; $@
</code></pre>



<p>% is a wild card character that will stand in for sample1, sample2, etc for the bam and fastq files 
$^ refers to your dependencies (%.unassembled.forward.fastq /%.unassembled.reverse.fastq)
$@ refers to your target (%.bam)</p>

<p>To process the make file you will type</p>



<pre><code>$make all
</code></pre>



<p>All together, Make will go through your list of output bams and process them all one by one (though it it possible to parallelize the process).
If you type make again, Make won't run again if it has processed all targets and rules. However, if you add in a new set of fastq files and run make, Make will just process the newly added fastq files. </p>



<pre><code># make a list of all the sam file targets you want to produce and define it as a variable 
DATABASE = output_files.txt 

# define the path to your reference genome as a variable 
REF= /path/to/ref/genome/hg38.fa

#define a variable that will act as a shell command that will read your list of output files via the shell cat command 
OUTPUTFILES := $(shell cat $(DATABASE))

all: $(OUTPUTFILES)

%.sam  :  %.unassembled.forward.fastq /%.unassembled.reverse.fastq
bwa mem -t 10 $(REF) $^ &gt; $@
</code></pre>



<p>If you have multiple target rules, and you only want to run a rule for a specific target, write</p>



<pre><code>$make &lt;target_name&gt;
</code></pre>



<p>The complete code </p>



<pre><code>## 3. Example 2  (More Advanced)

* Task:
* Download a series of cram files for the EPAS1 gene (chr2:46293667-46386703) from the 1000 Genomes database from a number of different individuals 
* These cram files are already aligned to the human reference genome. 
* We will merge these files and they can be used for functional or population genomics analyses. 

The command line task to obtain a cram file for a single sample is:
</code></pre>



<p>samtools view -C -h ftp:/ftp.1000genomes.ebi.ac.uk/vol1/ftp/data<em>collections/1000</em>genomes<em>project/data/CHS/HG00419/high</em>cov<em>alignment/HG00419.alt</em>bwamem<em>GRCh38DH.20150917.CHS.high</em>coverage.cram chr2:46293667-46386703 > EPAS1_indiv1.cram</p>



<pre><code>Samtools is software for manipulating alignments of short genomic reads to a reference genome as either sam/bam/cram. Cram is a very compressed version of the file and -C indicates our preference for this format. -h indicates to include the "header" which is a lot of metadata before the actual alignments that may include the sequencing center, parts of the reference genome, samples, etc 

The command line for several samples is
</code></pre>



<p>samtools view -C -h ftp:/ftp.1000genomes.ebi.ac.uk/vol1/ftp/data<em>collections/1000</em>genomes<em>project/data/CHS/HG00419/high</em>cov<em>alignment/HG00419.alt</em>bwamem<em>GRCh38DH.20150917.CHS.high</em>coverage.cram chr2:46293667-46386703 > EPAS1<em>indiv1.cram
samtools view -C -h ftp:/ftp.1000genomes.ebi.ac.uk/vol1/ftp/data</em>collections/1000<em>genomes</em>project/data/GIH/NA20845/high<em>cov</em>alignment/NA20845.alt<em>bwamem</em>GRCh38DH.20150917.GIH.high<em>coverage.cram chr2:46293667-46386703> EPAS1</em>indiv2.cram
samtools view -C -h ftp:/ftp.1000genomes.ebi.ac.uk/vol1/ftp/data<em>collections/1000</em>genomes<em>project/data/ASW/NA19625/high</em>cov<em>alignment/NA19625.alt</em>bwamem<em>GRCh38DH.20150917.ASW.high</em>coverage.cram chr2:46293667-46386703 > EPAS1_indiv3.cram</p>



<pre><code>What is you had 10s or 100s files to process? This is where GNu Make can help.
</code></pre>



<h1>list of 1000 Genome individuals from which we will grab data</h1>

<p>input = /Users/arcova/Documents/Service/Bioinformatics<em>AAAGs/Makefiles/sample</em>list_short.txt #the := will expand only once whereas = will </p>

<h1>define region of interest for EPAS1 gene, location obtained from ensembl.org for GRCh38.p12 human reference genome</h1>

<p>REGION= chr2:46293667-46386703 </p>

<h1>list of output crams</h1>

<p>output := /Users/arcova/Documents/Service/Bioinformatics<em>AAAGs/Makefiles/output</em>files_short.txt
OUTPUTFILES:= $(shell cat $(output))</p>

<p>.PHONY: all</p>

<p>all: $(OUTPUTFILES) rename</p>

<h1>get EPAS1 sequences from 1000 Genome dataset</h1>

<p>%.cram : 
    @echo '<strong><em>*' Extracting reads for $@ ...
    samtools view -C -h <code>grep $@ $(input)</code> $(REGION) > $@
    @echo '</em>*</strong>' Success</p>

<h1>rename cram files to reflect that it is part of the EPAS1 gene</h1>

<p>rename:
    for file in *.cram ; do \
    mv $file ${file//high_coverage.cram/EPAS1.cram} \
    done</p>

<p>```</p>

<h2>Resources and References for this Presentation</h2>

<p>https://v4.software-carpentry.org/make/index.html</p>

<p>Greg Wilson: "Software Carpentry: Getting Scientists to Write Better
Code by Making Them More Productive".  Computing in Science &amp;
Engineering, Nov-Dec 2006.</p>

<p>https://www.gnu.org/software/make/</p>
